src/com/ms/silverking/collection/MapUtil.java

```diff
@@ -12,72 +12,77 @@
-                    default: throw new RuntimeException("Panic");
-                    }
-                }
-            }
-        } while (line != null);
-        return map;
-    }
-    
-    public static <K,V> Pair<K,V> mapEntryToPair(Map.Entry<K,V> entry) {
-        return new Pair<>(entry.getKey(), entry.getValue());
-    }
-    
-    private static class StringParser implements Function<String,String> {
-        @Override
-        public String apply(String s) {
-            return s;
+  public enum NoDelimiterAction {Ignore, Warn, Exception}
+
+  ;
+
+  public static Map<String, String> parseStringMap(InputStream in, char delimiter, NoDelimiterAction noDelimiterAction)
+      throws IOException {
+    return parseMap(in, delimiter, noDelimiterAction, new StringParser(), new StringParser());
+  }
+
+  public static <K, V> Map<K, V> parseMap(InputStream in, char delimiter, NoDelimiterAction noDelimiterAction,
+      Function<String, K> keyParser, Function<String, V> valueParser) throws IOException {
+    BufferedReader reader;
+    String line;
+    int lineNumber;
+    Map<K, V> map;
+
+    map = new HashMap();
+    lineNumber = 0;
+    reader = new BufferedReader(new InputStreamReader(in));
+    do {
+      ++lineNumber;
+      line = reader.readLine();
+      if (line != null) {
+        int dIndex;
+
+        dIndex = line.indexOf(delimiter);
+        if (dIndex >= 0) {
+          String kDef;
+          String vDef;
+
+          kDef = line.substring(0, dIndex);
+          vDef = line.substring(dIndex + 1);
+          map.put(keyParser.apply(kDef), valueParser.apply(vDef));
+        } else {
+          switch (noDelimiterAction) {
+          case Ignore:
+            break;
+          case Warn:
+            Log.warningf("No delimiter found on line %d", lineNumber);
+            break;
+          case Exception:
+            throw new RuntimeException("No delimiter found on line: " + lineNumber);
+          default:
+            throw new RuntimeException("Panic");
+          }
         }
+      }
+    } while (line != null);
+    return map;
+  }
+
+  public static <K, V> Pair<K, V> mapEntryToPair(Map.Entry<K, V> entry) {
+    return new Pair<>(entry.getKey(), entry.getValue());
+  }
+
+  private static class StringParser implements Function<String, String> {
+    @Override
+    public String apply(String s) {
+      return s;
     }
-    
-    public static void main(String[] args) {
-        try {
-            Map<String,String>    m;
-            
-            m = MapUtil.parseStringMap(new FileInputStream(args[0]), '\t', MapUtil.NoDelimiterAction.Warn);
-            m.put("1","1");
-            System.out.println(m);
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
+  }
+
+  public static void main(String[] args) {
+    try {
+      Map<String, String> m;
+
+      m = MapUtil.parseStringMap(new FileInputStream(args[0]), '\t', MapUtil.NoDelimiterAction.Warn);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).